### ISSUE TITLE
magit-completing-read is not used everywhere one might expect.

### ISSUE THREAD

**Issue by tee3** [magit/magit#2668]
I really like the simplicity ido and the fact that it is an official part of Emacs, so I tried it out with Magit recently.  It works very well and provides the same interface as expected from `ido`.

However, I found a few places `magit-completing-read` is not used, which ends up not using `ido` (or anything else one might use for `magit-completing-read-function`.

Here is a partial list of the commands that do not seem to end using `magit-completing-read`

- [ ] `magit-log` - when using `l o`
- [ ] `magit-diff`- when using 'd r'
- [ ] `magit-format-patch` - when using `W p`
- [ ] `magit-ediff-compare` - when using `E r`

On a side note, I kind of expected `ido-everywhere` to be honored, so I enable it via a the `magit-mode-hook`.  Do you think this would be a good default?

```elisp
(defun my-magit-choose-completing-read-function ()
  (if ido-everywhere
    (setq magit-completing-read-function 'magit-ido-completing-read)))
```

**Comment by npostavs** [comment]
> Here is a partial list of the commands that do not seem to end using `magit-completing-read`

All those cases read more than one value which needs a different interface that ido isn't able to handle.

**Comment by tarsius** [comment]
> Do you think this would be a good default?

No because the "everywhere" in `ido-everywhere` is far from the truth anyway. (I recommend you also use `ido-ubiquitous`.)

**Comment by tee3** [comment]
Thanks for the feedback.

**Comment by ylluminarious** [comment]
I noticed the same commands not working with `ido` in `magit` even after setting `magit-completing-read-function` to `magit-ido-completing-read`. It would be quite nice to have `ido` support for these commands and others like them or at the very least for there to be a workaround that can make these commands work with `ido`. I'm also using `ido-ubiquitous`, but it does nothing to add `ido` support to these commands even if you add, say, `magit-log` to the `ido-ubiquitous-default-command-overrides` variable.

**Comment by jaccarmac** [comment]
This seems to be (recently-ish?) broken. I noticed it a long time ago in `magit-log-other`, assumed it was user error, and only just tracked down this issue. The code in `magit-utils` doesn't seem to be bit-rotten; Is there a particular version dependency w.r.t. `ido-completing-read+`, or a regression there? Is anyone else using `ido` for these commands any more?

**Comment by tarsius** [comment]
I would look on reddit for someone else who still uses ido.

**Comment by jaccarmac** [comment]
I haven't poked anyone else yet, but I did poke around my Emacs config and figure out what I think the problem is. It's in `crm-custom-mode`, and (I think) with the advice that mode puts on `completing-read-multiple`. I wasn't able to fully understand my debugging session, but with `crm-custom-mode` on, `magit-completing-read-multiple*` returns `nil`, which causes the a `split-string` call to explode in `magit-log-other`. `crm-custom-mode` contains a warning in the docs: "Note that `crm-separator` is purely cosmetic when this mode is enabled. It cannot actually be used as a separator." It looks like `magit-completing-read-multiple*` uses the separator at points.

**Comment by tarsius** [comment]
https://github.com/DarwinAwardWinner/crm-custom

**Comment by tarsius** [comment]
@DarwinAwardWinner is there any additional setup that allows using your `crm-custom-mode` with Magit or can they not be used together?

**Comment by DarwinAwardWinner** [comment]
Hmm, AFAIK crm-custom-mode should be fully compatible with any reasonable use of `completing-read-multiple`, and `magit-completing-read-multiple*` seems to be using it reasonably, so in theory it should work. If someone can give me a short code snippet that reproduces the error, I can try to debug it.

Aside from that, crm-custom is considered experimental and has significant caveats even when working properly, so I don't recommend using it unless you understand and accept those caveats.

**Comment by DarwinAwardWinner** [comment]
In fact, if you do have a reproducible example, please open an issue [here](https://github.com/DarwinAwardWinner/crm-custom/issues)

**Comment by jaccarmac** [comment]
All right, I have a minimal example which reproduces, but it's truly bizarre. I can open an issue if you'd like, but this may have something to do with the peculiarities of my environment. In an `emacs -Q`:

```elisp
(package-initialize)
(require 's)
(require 'magit)
(require 'crm-custom)
(crm-custom-mode)
(magit-completing-read-multiple* "" nil nil nil nil nil nil nil t)
```

The `(require 's)` is required, otherwise it works in a clean Emacs. I figured it out after crashing through the debugger trying to figure out why that clean version worked. The execution of `magit-completing-read-multiple*` differed during the advice, and its call to `s-split`. So in the clean Emacs, I tried to `C-h f` that function, failing until I had evaluated the `(package-initialize)` form.
